src/blocks/IpaPcs.sol

```diff
@@ -0,0 +1,324 @@
+
+        msm_index = 0;
+        Grumpkin.GrumpkinAffinePoint[] memory bases = new Grumpkin.GrumpkinAffinePoint[](msm_len);
+        for (uint256 index = 0; index < input.L_vec.length; index++) {
+            bases[msm_index] = input.L_vec[index];
+            msm_index++;
+        }
+        for (uint256 index = 0; index < input.R_vec.length; index++) {
+            bases[msm_index] = input.R_vec[index];
+            msm_index++;
+        }
+        bases[msm_index] = P;
+
+        return Grumpkin.multiScalarMul(bases, scalars);
+    }
+
+    function compute_P_hat_right(
+        uint256 b_hat,
+        uint256 a_hat,
+        Grumpkin.GrumpkinAffinePoint memory ck_hat,
+        Grumpkin.GrumpkinAffinePoint memory ck_c
+    ) private returns (Grumpkin.GrumpkinAffinePoint memory) {
+        Grumpkin.GrumpkinAffinePoint[] memory bases = new Grumpkin.GrumpkinAffinePoint[](2);
+        bases[0] = ck_hat;
+        bases[1] = ck_c;
+
+        uint256[] memory scalars = new uint256[](2);
+        scalars[0] = a_hat;
+        scalars[1] = mulmod(a_hat, b_hat, Grumpkin.P_MOD);
+
+        return Grumpkin.multiScalarMul(bases, scalars);
+    }
+
+    function verifyGrumpkin(IpaInputGrumpkin memory input, KeccakTranscriptLib.KeccakTranscript memory transcript)
+        public
+        returns (bool)
+    {
+        uint256 n = 2 ** input.point.length;
+
+        uint256[] memory b_vec = EqPolynomialLib.evals(input.point, Grumpkin.P_MOD, Grumpkin.negateBase);
+        (Grumpkin.GrumpkinAffinePoint[] memory ck1,) = split_at(input.ck_v, b_vec.length);
+
+        // b"IPA" in Rust
+        uint8[] memory label = new uint8[](3);
+        label[0] = 0x49;
+        label[1] = 0x50;
+        label[2] = 0x41;
+
+        transcript = KeccakTranscriptLib.dom_sep(transcript, label);
+
+        if (b_vec.length != n) {
+            revert("NovaError::InvalidInputLength");
+        }
+        if (n != 1 << input.L_vec.length) {
+            revert("NovaError::InvalidInputLength");
+        }
+        if (input.L_vec.length != input.R_vec.length) {
+            revert("NovaError::InvalidInputLength");
+        }
+        if (input.L_vec.length >= 32) {
+            revert("NovaError::InvalidInputLength");
+        }
+
+        // b"U" in Rust
+        label = new uint8[](1);
+        label[0] = 0x55;
+
+        transcript =
+            KeccakTranscriptLib.absorb(transcript, label, InstanceGrumpkin(input.commitment, b_vec, input.eval));
+
+        // b"r" in Rust
+        label = new uint8[](1);
+        label[0] = 0x72;
+        uint256 r;
+        (transcript, r) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curveGrumpkin(), label);
+
+        uint256[] memory r_vec = new uint256[](input.L_vec.length);
+        for (uint256 index = 0; index < r_vec.length; index++) {
+            // b"L" in Rust
+            label[0] = 0x4c;
+            transcript = KeccakTranscriptLib.absorb(transcript, label, input.L_vec[index].x);
```

Comment:
So in `arecibo`, I could find the implementation of `Keccak256Transcript::absorb` and the use of  `to_transcript_bytes` but I could not find the implementation for the latter for `CompressedCommitment`.

Why is it that we only need to absorb the `x` of the curve point and not `x`+`y`+`infinity_byte` as we are in a lot of our implementations in arecibo?

Reply:
> The implementation of to_transcript_bytes for CompressedCommitment is indeed hidden in [following](https://github.com/lurk-lab/arecibo/blob/dev/src/provider/traits.rs#L161) macro.

Thanks! I understand now that it trace all the way to the defined `Compressed` type which in our case is the `G1Compressed` structure defined through the usage of the `new_curve_impl` macro in the `halo2` crate.

> As far as I know, all the necessary information about y coordinate is already stored in x.

So from what I get of your sentence and a few research, the compressed point is the data of `x` along with a sign bit that is derived from the `y` value of an affine point. And it seems that this is what we absorb in the Rust implementation. Aren't we missing the sign bit then? Or do we not need it?

Reply:
I think sign is also encoded in the information behind `x` (consider [from_bytes](https://github.com/privacy-scaling-explorations/halo2curves/blob/main/src/derive/curve.rs#L106) function in halo2curves). The field element for Bn256 (and Grumpkin) have 254 bits in size ([link](https://hackmd.io/@aztec-network/ByzgNxBfd#2-Grumpkin---A-curve-on-top-of-BN-254-for-SNARK-efficient-group-operations)), while `x` is `uint256`, which is 256 bits - so we have two bits for the metadata and by absorbing whole `uint256` we absorb sign as well. 

For the uncompressed commitment (which is essentially a Grumpkin affine point with two coordinates) we have different [absorb](https://github.com/lurk-lab/solidity-verifier/blob/main/src/blocks/KeccakTranscript.sol#L1235) where we put `x`, `y` and an extra-byte representing whether point is infinity (usually this is not the case, so this is 0x00) to the transcript's memory.

Reply:
I believe the authoritative format for this conversion is in [SECG1](https://www.secg.org/sec1-v2.pdf), section 2.3.3. read with compression active.